examples/experimental/bert_squad_pytorch/model_def.py

```diff
@@ -0,0 +1,166 @@
+            self.context.get_hparam("model_type")
+        ]
+        self.validation_dataset, self.examples, self.features = data.load_and_cache_examples(
+            data_dir=self.download_directory,
+            config=self.context.get_data_config(),
+            model_type=self.context.get_hparam("model_type"),
+            max_seq_length=self.context.get_hparam("max_seq_length"),
+            doc_stride=self.context.get_hparam("doc_stride"),
+            max_query_length=self.context.get_hparam("max_query_length"),
+            model_name_or_path=self.context.get_data_config().get("model_name_or_path"),
+            evaluate=True,
+            output_examples=True,
+        )
+        self.tr_loss = 0.0
+
+    def build_training_data_loader(self) -> DataLoader:
+        train_dataset = data.load_and_cache_examples(
+            data_dir=self.download_directory,
+            config=self.context.get_data_config(),
+            model_type=self.context.get_hparam("model_type"),
+            max_seq_length=self.context.get_hparam("max_seq_length"),
+            doc_stride=self.context.get_hparam("doc_stride"),
+            max_query_length=self.context.get_hparam("max_query_length"),
+            model_name_or_path=self.context.get_data_config().get("model_name_or_path"),
+            evaluate=False,
+            output_examples=False,
+        )
+        return DataLoader(train_dataset, batch_size=self.context.get_per_slot_batch_size())
+
+    def build_validation_data_loader(self) -> DataLoader:
+        return DataLoader(
+            self.validation_dataset,
+            batch_size=self.context.get_per_slot_batch_size(),
+        )
+
+    def build_model(self) -> torch.nn.modules.module.Module:
+        cache_dir_per_rank = f"/tmp/{self.context.distributed.get_rank()}"
+
+        config = self.config_class.from_pretrained(
+            self.context.get_data_config().get("model_name_or_path"),
+            cache_dir=cache_dir_per_rank,
+        )
+        model = self.model_class.from_pretrained(
+            self.context.get_data_config().get("model_name_or_path"),
+            from_tf=bool(".ckpt" in self.context.get_data_config().get("model_name_or_path")),
+            config=config,
+            cache_dir=cache_dir_per_rank,
+        )
+        return model
+
+    def optimizer(self, model: nn.Module) -> torch.optim.optimizer.Optimizer:
+        no_decay = ["bias", "LayerNorm.weight"]
+        optimizer_grouped_parameters = [
+            {
+                "params": [
+                    p for n, p in model.named_parameters() if not any(nd in n for nd in no_decay)
+                ],
+                "weight_decay": self.context.get_hparam("weight_decay"),
+            },
+            {
+                "params": [
+                    p for n, p in model.named_parameters() if any(nd in n for nd in no_decay)
+                ],
+                "weight_decay": self.context.get_hparam("weight_decay"),
+            },
```

Comment:
blocking: I'm not sure I understand why this first argument to `AdamW` needs two items in the list. 

I looked at the original `run_squad` example also don't understand it, although it looks as though it uses different values for the `weight_decay` key (first using an arg, then using 0.0). Can you investigate what's going on here?

Reply:
Looking at link pointed to by this: https://github.com/huggingface/transformers/issues/492. In Lines 130-150 in the file pointed to by the link above, they seem to be modifying the OG Adam optimizer to handle L2 regularization in a particular way (while maintaining the original functionality with no weight decay). Essentially, they seem to be separating the weight decay out per parameter. So I believe in my code, Line 92 here should be setting `"weight_decay": 0.0` per what is in `run_squad`